bbot/core/helpers/asn.py

```diff
@@ -0,0 +1,204 @@
+                "name": raw.get("asn_name", ""),
+                "description": raw.get("org", ""),
+                "country": raw.get("country", ""),
+            }
+            return [rec]
+
+        log.warning(f"ASN DB API: returned unexpected format for {ip}: {raw}")
+        return None
+
+    async def _query_api_asn(self, asn: str):
+        url = f"{self.asndb_asn_url}{asn}"
+        response = await self._request_with_retry(url)
+        if response is None:
+            log.warning(f"ASN DB API: no response for {asn}")
+            return None
+
+        status = getattr(response, "status_code", 0)
+        if status != 200:
+            log.warning(f"ASN DB API: returned {status} for {asn}")
+            return None
+
+        try:
+            raw = response.json()
+        except Exception as e:
+            log.warning(f"ASN DB API: JSON decode error for {asn}: {e}")
+            return None
+
+        if isinstance(raw, dict):
+            subnets = raw.get("subnets")
+            if isinstance(subnets, str):
+                subnets = [subnets]
+            if not subnets:
+                subnets = []
+
+            rec = {
+                "asn": str(raw.get("asn", "")),
+                "subnets": subnets,
+                "name": raw.get("asn_name", ""),
+                "description": raw.get("org", ""),
+                "country": raw.get("country", ""),
+            }
+            return [rec]
+
+        log.warning(f"ASN DB API: returned unexpected format for {asn}: {raw}")
+        return None
+
+    def _cache_store_asn(self, asn_list, asn_id: int):
+        """Cache ASN data by ASN ID"""
+        self._asn_to_data_cache[asn_id] = asn_list
+        log.debug(f"ASN cache ADD {asn_id} -> {asn_list[0].get('asn', '?') if asn_list else '?'}")
+
+    def _cache_lookup_asn(self, asn_id: int):
+        """Lookup cached ASN data by ASN ID"""
+        return self._asn_to_data_cache.get(asn_id)
+
+    def _cache_store_ip(self, asn_list):
+        if not (self._tree4 or self._tree6):
+            return
+        for rec in asn_list:
+            subnets = rec.get("subnets") or []
+            if isinstance(subnets, str):
+                subnets = [subnets]
+            for p in subnets:
+                try:
+                    net = ipaddress.ip_network(p, strict=False)
+                except ValueError:
+                    continue
+                tree = self._tree4 if net.version == 4 else self._tree6
+                tree.insert(str(net), data=asn_list)
+                self._subnet_to_asn_cache[str(net)] = asn_list
+                log.debug(f"IP cache ADD {net} -> {asn_list[:1][0].get('asn', '?')}")
+
+    def _cache_lookup_ip(self, ip: str):
+        ip_obj = ipaddress.ip_address(ip)
+        tree = self._tree4 if ip_obj.version == 4 else self._tree6
+        node = tree.get_node(ip)
+        if node and getattr(node, "data", None):
+            return node.data
+        return None
```

Comment:
We should cache negative/nonexistent IPs + ASNs and distinguish between non-cached and non-existent to avoid making unnecessary requests

Reply:
fixed in https://github.com/blacklanternsecurity/bbot/pull/2553/commits/bdce02d64b737c9f0db2b289b46b4b696c1b4d48